internal/infrastructure/mongo/mongodoc/client.go

```diff
@@ -202,45 +202,70 @@ func getCursor(raw bson.Raw, key string) (*usecase.Cursor, error) {
 	return &c, nil
 }
 
-func (c *Client) Paginate(ctx context.Context, col string, filter interface{}, p *usecase.Pagination, consumer Consumer) (*usecase.PageInfo, error) {
+func (c *Client) Paginate(ctx context.Context, col string, filter bson.M, sort *string, p *Pagination, consumer Consumer) (*usecase.PageInfo, error) {
 	if p == nil {
 		return nil, nil
 	}
 	coll := c.Collection(col)
+	const key = "id"
 
-	key := "id"
+	findOptions := options.Find()
+	findOptions.SetCollation(&options.Collation{Strength: 1, Locale: "en"})
+
+	var sortOptions bson.D
+	var sortKey = ""
+	if sort != nil && len(*sort) > 0 && *sort != "id" {
+		sortKey = *sort
+		sortOptions = append(sortOptions, bson.E{Key: sortKey, Value: p.SortDirection()})
+	}
+	sortOptions = append(sortOptions, bson.E{Key: key, Value: p.SortDirection()})
+
+	findOptions.Sort = sortOptions
 
 	count, err := coll.CountDocuments(ctx, filter)
 	if err != nil {
 		return nil, fmt.Errorf("failed to count documents: %v", err.Error())
 	}
 
-	reverse := false
-	var limit int64
-	findOptions := options.Find()
-	if first := p.First; first != nil {
-		limit = int64(*first)
-		findOptions.Sort = bson.D{
-			{Key: key, Value: 1},
-		}
-		if after := p.After; after != nil {
-			filter = appendE(filter, bson.E{Key: key, Value: bson.D{
-				{Key: "$gt", Value: *after},
-			}})
-		}
+	limit, op, cur, err := p.Parameters()
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse pagination parameters: %w", err)
 	}
-	if last := p.Last; last != nil {
-		reverse = true
-		limit = int64(*last)
-		findOptions.Sort = bson.D{
-			{Key: key, Value: -1},
+
+	if cur != nil {
+
+		var paginationFilter bson.M
+
+		if sortKey == "" {
+			paginationFilter = bson.M{key: bson.M{op: *cur}}
+		} else {
+			var curObj bson.M
+			if err := coll.FindOne(ctx, bson.M{key: *cur}).Decode(&curObj); err != nil {
+				return nil, fmt.Errorf("failed to find cursor")
+			}
+			if curObj[sortKey] == nil {
+				return nil, fmt.Errorf("invalied sort key")
+			}
+			paginationFilter = bson.M{
+				"$or": []bson.M{
+					{sortKey: bson.M{"$gt": curObj[sortKey]}},
+					{
+						sortKey: curObj[sortKey],
+						key:     bson.M{"$gt": *cur},
+					},
+				},
+			}
 		}
-		if before := p.Before; before != nil {
-			filter = appendE(filter, bson.E{Key: key, Value: bson.D{
-				{Key: "$lt", Value: *before},
-			}})
+
```

Comment:
Processing to build paginationFilter is very long, so separate it to another function. It is hard to read codes